src/Microsoft.Health.Dicom.Core/Features/Workitem/WorkitemService.ChangeState.cs

```diff
@@ -0,0 +1,113 @@
+        CancellationToken cancellationToken = default)
+    {
+        EnsureArg.IsNotNull(dataset, nameof(dataset));
+
+        try
+        {
+            var workitemMetadata = await _workitemOrchestrator
+                .GetWorkitemMetadataAsync(workitemInstanceUid, cancellationToken)
+                .ConfigureAwait(false);
+            if (workitemMetadata == null)
+            {
+                _responseBuilder.AddFailure(
+                    FailureReasonCodes.UpsInstanceNotFound,
+                    DicomCoreResource.WorkitemInstanceNotFound,
+                    dataset);
+
+                return _responseBuilder.BuildChangeWorkitemStateResponse();
+            }
+
+            ValidateChangeWorkitemStateRequest(dataset, workitemMetadata);
+
+            // TODO: Update TransactionUID in SQL
+
+            var originalBlobDicomDataset = await _workitemOrchestrator
+                .GetWorkitemBlobAsync(workitemMetadata, cancellationToken)
+                .ConfigureAwait(false);
+
+            var updateDataset = PrepareChangeWorkitemStateDicomDataset(originalBlobDicomDataset, dataset);
+
+            await _workitemOrchestrator.UpdateWorkitemStateAsync(
+                    updateDataset,
+                    workitemMetadata,
+                    dataset.GetProcedureStepState(),
+                    cancellationToken)
+                .ConfigureAwait(false);
+
+            _responseBuilder.AddSuccess(string.Empty);
+        }
+        catch (Exception ex)
+        {
+            ushort? failureCode = FailureReasonCodes.ProcessingFailure;
+            switch (ex)
+            {
+                case BadRequestException:
+                case DicomValidationException:
+                case DatasetValidationException:
+                case ValidationException:
+                    failureCode = FailureReasonCodes.ValidationFailure;
+                    break;
+
+                case WorkitemUpdateNotAllowedException:
+                    failureCode = FailureReasonCodes.UpsInstanceUpdateNotAllowed;
+                    break;
+
+                case ItemNotFoundException:
+                    failureCode = FailureReasonCodes.UpsInstanceNotFound;
+                    break;
+            }
+
+            _logger.LogInformation(ex,
+                "Change workitem state failed for the DICOM instance work-item entry. Failure code: {FailureCode}.", failureCode);
+
+            _responseBuilder.AddFailure(failureCode, ex.Message);
+        }
+
+        return _responseBuilder.BuildChangeWorkitemStateResponse();
+    }
+
+    private static DicomDataset PrepareChangeWorkitemStateDicomDataset(DicomDataset dataset, DicomDataset originalBlobDicomDataset)
+    {
+        var resultDataset = originalBlobDicomDataset
+            .AddOrUpdate(DicomTag.TransactionUID, dataset.GetString(DicomTag.TransactionUID))
+            .AddOrUpdate(DicomTag.ProcedureStepState, dataset.GetString(DicomTag.ProcedureStepState));
+
```

Comment:
If its a cancel, should we update cancelation reason. Do we need to call CAncel method

Reply:
Done